dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,154 @@
+		getConfigConn: fn,
+		workerName:    workerName,
+	}
+}
+
+func (c *connAmountChecker) check(ctx context.Context, checkerName string) *Result {
+	result := &Result{
+		Name:  checkerName,
+		Desc:  "check if user-specified amount of connection exceeds downstream's maximum",
+		State: StateFailure,
+	}
+	baseConn, err := c.targetDB.GetBaseConn(ctx)
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	var rows *sql.Rows
+	rows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW GLOBAL VARIABLES LIKE 'max_connections'")
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	defer rows.Close()
+	var (
+		maxConn   int
+		_variable string
+	)
+	for rows.Next() {
+		err = rows.Scan(&_variable, &maxConn)
+		if err != nil {
+			markCheckError(result, err)
+			return result
+		}
+	}
+	neededConn := c.getConfigConn(c.stCfgs)
+	if neededConn > maxConn {
+		result.Errors = append(
+			result.Errors,
+			NewError("downstream's max_connections: %d is less than the amount %s needs: %d", maxConn, c.workerName, neededConn),
+		)
+		result.Instruction = "set larger max_connections or reduce the pool size of dm"
+		result.State = StateFailure
+	} else {
+		result.State = StateSuccess
+	}
+	return result
+}
+
+type SyncerConnAmountChecker struct {
+	connAmountChecker
+}
+
+func NewSyncerConnAmountCheker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig) RealChecker {
+	return &SyncerConnAmountChecker{
+		connAmountChecker: newConnAmountChecker(targetDB, stCfgs, func(stCfgs []*config.SubTaskConfig) int {
+			syncerConn := 0
+			for _, stCfg := range stCfgs {
+				// syncer's worker and checkpoint (always keeps one db connection)
+				syncerConn += stCfg.SyncerConfig.WorkerCount + 1
+			}
+			return syncerConn
+		}, "syncer"),
+	}
+}
+
+func (s *SyncerConnAmountChecker) Name() string {
+	return "syncer_conn_amount_checker"
+}
+
+func (s *SyncerConnAmountChecker) Check(ctx context.Context) *Result {
+	return s.check(ctx, s.Name())
+}
+
+type LoaderConnAmountChecker struct {
+	connAmountChecker
+}
+
+func NewLoaderConnAmountChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig) RealChecker {
+	return &LoaderConnAmountChecker{
+		connAmountChecker: newConnAmountChecker(targetDB, stCfgs, func(stCfgs []*config.SubTaskConfig) int {
+			loaderConn := 0
+			for _, stCfg := range stCfgs {
+				// loader's worker and checkpoint (always keeps one db connection)
+				loaderConn += stCfg.LoaderConfig.PoolSize + 1
+			}
+			return loaderConn
+		}, "loader"),
+	}
+}
+
+func (l *LoaderConnAmountChecker) Name() string {
+	return "loader_conn_amount_checker"
+}
+
+func (l *LoaderConnAmountChecker) Check(ctx context.Context) *Result {
+	return l.check(ctx, l.Name())
+}
+
+func NewDumperConnAmountChecker(targetDB *conn.BaseDB, dumperThreads int) RealChecker {
+	return &DumperConnAmountChecker{
+		connAmountChecker: newConnAmountChecker(targetDB, nil, func(_ []*config.SubTaskConfig) int {
+			return dumperThreads
```

Comment:
Dumpling needs a connection for generating dumping SQL and a connection for consistency control.
```suggestion
			return dumperThreads+2
```